src/librustc_typeck/check/mod.rs

```diff
@@ -3764,7 +3766,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                           coerce.coerce(self, &cause, e, e_ty);
                       } else {
                           assert!(e_ty.is_nil());
-                          coerce.coerce_forced_unit(self, &cause, &mut |_| (), true);
+                          // If we're breaking without a value, then the break
+                          // implicitly carries a `()` value. Unless, that is,
+                          // the function is already diverging. In this case, we
+                          // don't need to supply any value (in which case, the
+                          // value will be considered to have type `!`).
+                          if !prev_diverges.always() {
```

Comment:
Hmm, I think if we are going to "do this", we might want to consider "doing it right". In particular, the value of the diverges flag on entry to an expression does not indicate if "all prior points in the CFG diverged" but rather if *some preceding subexpression within the parent diverged*. 

This means that while this test works now:

```rust
    if false {
        let _: ! = {
            loop { return; break } // ok
        };
}
```

I expect that this one will fail, right?

```rust
    if false {
        let _: ! = {
            loop { 
                return; 
                { break; } // now the break is part of a sub-block...
            }
        };
}
```

Put another way, I am wary of adding rules that try to use the existing divergence mechanism to figure out if **prior code** had diverged.

Reply:
Yes, that's correct.